frontend/device/android/powerd.lua

```diff
@@ -68,20 +64,45 @@ function AndroidPowerD:turnOffFrontlightHW()
     end
     android.setScreenBrightness(self.fl_min)
     self.is_fl_on = false
-    broadcastLightChanges()
+
+    self:broadcastLightChanges()
 end
 
 function AndroidPowerD:turnOnFrontlightHW()
+    if not self.is_fl_sw_on then -- on devices with a software frontlight switch (e.g Tolinos), enable it
+        android:enableFrontlightSwitch()
+        self.is_fl_sw_on = true
+    end
+
     if self:isFrontlightOn() and self:isFrontlightOnHW() then
         return
     end
-    -- on devices with a software frontlight switch (e.g Tolinos), enable it
-    android.enableFrontlightSwitch()
 
     android.setScreenBrightness(math.floor(self.fl_intensity * self.bright_diff / self.fl_max))
-
     self.is_fl_on = true
-    broadcastLightChanges()
+
+    self:broadcastLightChanges()
+end
+
+function AndroidPowerD:getFrontlightSwitchState()
+    return self.is_fl_sw_on
+end
+
+function AndroidPowerD:detectedFrontlightSwitchToggle()
+    if self.is_fl_sw_on and not android:getFrontlightSwitchState() then
+        android:enableFrontlightSwitch() --this sends keypress to android (and KOReader)
```

Comment:
it looks like a loop

Reply:
Only at the first glance. 
Imagine:
1.) HardwareSwitch is on first -> is_fl_sw_on==true and getFrontligthSwitchState==true
2.) Someone toggle the Hardware switch -> is_fl_sw_on=true and getFrontlightSwitchState==false
3.) We send another keypress to the system (to toggle HardwareSwitch on again to on)
4.) This time is_fl_sw_on==true and getFrontlightSwitchState==true -> no keypress

Reply:
But `android.enableFrontlightSwitch()` (there's no need to pass self) requires root. Or not? what happens if it requires root and the device is not rooted?